File: final/q9.hs
shouldBeUnique :: Eq a => [a] -> [a] -> [a]
shouldBeUnique lis res
  | null lis = res
  | null res = shouldBeUnique (tail lis) [head lis]
  | head lis == head res = shouldBeUnique (tail lis) res
  | otherwise = shouldBeUnique (tail lis) (head lis : res)

unique :: Eq a => [a] -> [a]
unique a = reverse (shouldBeUnique a [])

File: final/q9.pl
shouldBeUnique([], [], []).
shouldBeUnique([H|AT], [], RealResult):- shouldBeUnique(AT, [H], RealResult).
shouldBeUnique([H|AT], Result, RealResult):- Result = [H|_], shouldBeUnique(AT, Result, RealResult).
shouldBeUnique([H|AT], Result, RealResult):- Result = [RH|_], not(RH is H), shouldBeUnique(AT, [H|Result], RealResult).
shouldBeUnique([], Result, RealResult):- RealResult = Result.

unique(A, Result):- shouldBeUnique(A, [], ResultR), reverse(ResultR, Result).